src/ReductionMsg.chpl

```diff
@@ -611,11 +619,40 @@ module ReductionMsg
       return res;
     }
     
-    proc segMean(values:[] ?t, segments:[?D] int): [D] real {
+    proc segMean(values:[] ?t, segments:[?D] int, skipNan=false): [D] real {
       var res: [D] real;
       if (D.size == 0) { return res; }
-      var sums = segSum(values, segments);
-      var counts = segCount(segments, values.size);
+      var sums;
+      var counts;
+      if (isFloatType(t) && skipNan) {
+        var arrCopy = makeDistArray(values.size, real);
+        var nancounts = makeDistArray(segments.size, int);
```

Comment:
I worry that the logic inside the `forall` here will be slow in multilocale. I might approach this with a more vectorized pattern, e.g.
```chapel
// count cumulative nans over all values
var cumnans = + scan isnan(values); // assuming isnan is promotable
// find cumulative nans at segment boundaries
var segnans: [segments.domain] int;
forall si in segments.domain {
    if si == segments.domain.high {
        segnans[si] = cumnans[cumnans.domain.high];
    } else {
        segnans[si] = cumnans[segments[si+1]-1];
    }
}
// take diffs of adjacent segments to find nan count in each segment
var nancounts = segnans[segnans.domain.interior(-(segnans.domain.size-1))] - segnans[segnans.domain.interior(segnans.domain.size-1)];
```

I have not checked the logic or syntax here (in particular, I always forget how `.interior` works), so treat this as pseudocode.

Reply:
Thanks for the suggestion, I think that your proposed implementation is much better, I added that in with some modifications.